### ISSUE TITLE
Question about the accuracy of FedFomo

### ISSUE THREAD

**Issue by Sherrylife** [TsingZ0/PFL-Non-IID#11]
Hi, TsingZ0.   
I'm sorry to disturb you again. This time, I would like to ask whether you can reproduce the experimental results of 15 clients and 100 clients in the CIFAR10 data set of Fedfomo paper. I ran through your code, and set the same hyperparameter  as in Fedfomo's original paper, using the CNN model similar to Fedavg's, it seems that the mean accuracy in Fedfomo's original paper cannot be achieved.

**Comment by TsingZ0** [comment]
Hi, Sherrylife. 

The Non-IID setup in this platform is not the same as in Fedfomo's original paper.

**Comment by TsingZ0** [comment]
Hi, @Sherrylife 

I have updated the code. Please check if this issue has been addressed.

**Comment by Sherrylife** [comment]
> Hi, @Sherrylife
> 
> I have updated the code. Please check if this issue has been addressed.

Hi TsingZ0, thank your for your reply. I compared the experimental effects of FedFomo and FedLocal using cifar10 dataset with **Practical non-iID** setting, and set *client fraction=1*, *learning rate=0.01*, *client number=15*, which are the same as the original paper of FedFomo . Surprisedly, I found that the accuracy rate of FedFomo training after 100 rounds was as high as 88%, while FedLocal was only 86%. FedFomo works better than FedLocal . I would like to ask where you have modified it to achieve this effect. Looking forward to your reply!

**Comment by TsingZ0** [comment]
Hi, @Sherrylife 
I only set `shuffle=True` in `./dataset/utils/dataset_utils.py`.

**Comment by Sherrylife** [comment]
> Hi, @Sherrylife I only set `shuffle=True` in `./dataset/utils/dataset_utils.py`.

Hi, @TsingZ0 , I have read your code for generating **Practical non-iid** setting. In fact, I think this method is not random enough, because the labels he assigns have fixed rules, as shown in the figure below(using my own code). For example, 0 always goes with 1, but not with other numbers.
![image](https://user-images.githubusercontent.com/60345931/176990577-de730d93-2b6d-469d-b3cd-e08d5d1c64c3.png)
Under this setting, **FedFomo**(88% accuracy) is better than **FedLocal**(86% accuracy). However, I only need to change the experimental settings **slightly**. For example, randomly assign data to each user so that they have no more than 2 tags of data, and the experimental results show that **FedFomo** and **FedLocal** have the same effect(both 87% accuracy), and there is no significant difference between them. I wonder how you only think about it?

**Comment by TsingZ0** [comment]
As mentioned in some papers, if the heterogeneity is too severe, local training without communication performs better.

**Comment by Sherrylife** [comment]
> 
Can you tell me exactly what the papers are?
Do you mean that if the heterogeneity is too severe, then local training is better than most of the personalized federated learning methods?
 My experience is that when there are few clients(**cross-silo**), there seem to be few personalized approaches that are better than native training. While when there are a large number of clients(**cross-device**), some personalized methods(e.g. **FedFomo**) can be better than local training.

**Comment by TsingZ0** [comment]
Sorry for forgetting the paper titles. 

> My experience is that when there are few clients(cross-silo), there seem to be few personalized methods that are better than native training. While when there are a large number of clients(cross-device), some personalized methods(e.g. FedFomo) can be better than local training.

The reason for this phenomenon is that the data amount on the clients in the cross-silo setting is enough for learning a reasonable local model, while the cross-device setting suffers from data shortage and requires federated learning.

**Comment by Sherrylife** [comment]
> Sorry for forgetting the paper titles.
> 
> > My experience is that when there are few clients(cross-silo), there seem to be few personalized methods that are better than native training. While when there are a large number of clients(cross-device), some personalized methods(e.g. FedFomo) can be better than local training.
> 
> The reason for this phenomenon is that the data amount on the clients in the cross-silo setting is enough for learning a reasonable local model, while the cross-device setting suffers from data shortage and requires federated learning.

Ok, thank you for your patience.
